File: src/devtwin/adapters/dotnet.py
""" ".NET ecosystem adapter (C#/F#/VB)."""

from __future__ import annotations

import json
from pathlib import Path

from devtwin.adapters.base import EcosystemAdapter, extract_version
from devtwin.core.models import DependencyInfo, HealthIssue, Presence, RuntimeInfo, Severity
from devtwin.core.runner import run_command, which

PROJECT_GLOBS = ["*.csproj", "*.fsproj", "*.vbproj", "**/*.csproj", "**/*.fsproj", "**/*.vbproj"]
SLN_GLOBS = ["*.sln"]


class DotnetAdapter(EcosystemAdapter):
    ecosystem = "dotnet"

    def detect(self, root: Path) -> bool:
        if (root / "global.json").exists():
            return True
        for pattern in [*PROJECT_GLOBS, *SLN_GLOBS]:
            if next(root.glob(pattern), None) is not None:
                return True
        return False

    def _required_version(self, root: Path) -> tuple[str | None, str | None]:
        global_json = root / "global.json"
        if not global_json.exists():
            return None, None
        try:
            data = json.loads(global_json.read_text(encoding="utf-8", errors="ignore"))
        except (OSError, json.JSONDecodeError):
            return None, None
        version = data.get("sdk", {}).get("version") if isinstance(data.get("sdk"), dict) else None
        return version, "global.json" if version else None

    def inspect_runtime(self, root: Path) -> list[RuntimeInfo]:
        required, source = self._required_version(root)
        dotnet_path = which("dotnet")
        if dotnet_path is None:
            return [
                RuntimeInfo(
                    name="dotnet",
                    presence=Presence.NOT_INSTALLED,
                    required_version=required,
                    source=source,
                )
            ]
        result = run_command([dotnet_path, "--version"], timeout=10)
        installed = extract_version(result.stdout) if result.available else None
        return [
            RuntimeInfo(
                name="dotnet",
                presence=Presence.INSTALLED if installed else Presence.UNKNOWN,
                installed_version=installed,
                required_version=required,
                path=dotnet_path,
                source=source,
            )
        ]

    def inspect_dependencies(self, root: Path) -> DependencyInfo | None:
        projects: list[str] = []
        for pattern in PROJECT_GLOBS:
            projects.extend(str(p.relative_to(root)) for p in root.glob(pattern))
        if not projects:
            return None
        return DependencyInfo(
            ecosystem="dotnet",
            manager="nuget",
            lockfile=None,
            manifest_present=True,
            notes=[f"{len(set(projects))} project file(s) found"],
        )

    def inspect_tests(self, root: Path) -> list[str]:
        return ["dotnet test"] if self.detect(root) else []

    def inspect_build_commands(self, root: Path) -> list[str]:
        return ["dotnet build"] if self.detect(root) else []

    def health_checks(self, root: Path, runtimes: list[RuntimeInfo]) -> list[HealthIssue]:
        issues = []
        for rt in runtimes:
            if rt.name == "dotnet" and rt.presence == Presence.NOT_INSTALLED:
                issues.append(
                    HealthIssue(
                        severity=Severity.CRITICAL,
                        code="dotnet.not_installed",
                        title=".NET SDK not found",
                        message="No `dotnet` executable was found on PATH.",
                        evidence=["which dotnet -> not found"],
                        recommendation="Install the .NET SDK from https://dotnet.microsoft.com/download.",
                    )
                )
        return issues

File: mcphub/devtwin/src/adapters/dotnet.ts
/**
 * .NET ecosystem adapter (C#/F#/VB).
 */

import { join } from 'node:path';

import {
  Presence,
  Severity,
  type DependencyInfo,
  type HealthIssue,
  type RuntimeInfo,
} from '../core/models.js';
import { runCommand, which } from '../core/runner.js';
import { globAny, pathExists, readJsonFile } from '../system/filesystem.js';
import { EcosystemAdapter, extractVersion } from './base.js';

const PROJECT_GLOBS = [
  '*.csproj',
  '*.fsproj',
  '*.vbproj',
  '**/*.csproj',
  '**/*.fsproj',
  '**/*.vbproj',
];
const SLN_GLOBS = ['*.sln'];

// Upper bound on how many project files we enumerate; the Python glob was
// unbounded but only the count and "did we find any" ever get used.
const MAX_PROJECT_MATCHES = 200;

export class DotnetAdapter extends EcosystemAdapter {
  readonly ecosystem = 'dotnet';

  detect(root: string): boolean {
    if (pathExists(join(root, 'global.json'))) {
      return true;
    }
    // One match from any project/solution glob is enough to claim the project.
    return globAny(root, [...PROJECT_GLOBS, ...SLN_GLOBS], 1).length > 0;
  }

  private requiredVersion(root: string): [string | null, string | null] {
    const globalJson = join(root, 'global.json');
    if (!pathExists(globalJson)) {
      return [null, null];
    }
    // readJsonFile degrades to null on unreadable *or* malformed JSON.
    const data = readJsonFile(globalJson);
    if (data === null) {
      return [null, null];
    }
    const sdk = (data as Record<string, unknown>)['sdk'];
    const version =
      sdk !== null && typeof sdk === 'object' && !Array.isArray(sdk)
        ? ((sdk as Record<string, unknown>)['version'] as string | undefined) ?? null
        : null;
    return [version, version ? 'global.json' : null];
  }

  override async inspectRuntime(root: string): Promise<RuntimeInfo[]> {
    const [required, source] = this.requiredVersion(root);
    const dotnetPath = which('dotnet');
    if (dotnetPath === null) {
      return [
        {
          name: 'dotnet',
          presence: Presence.NOT_INSTALLED,
          installed_version: null,
          required_version: required,
          path: null,
          source,
        },
      ];
    }
    const result = await runCommand([dotnetPath, '--version'], { timeout: 10 });
    const installed = result.available ? extractVersion(result.stdout) : null;
    return [
      {
        name: 'dotnet',
        presence: installed ? Presence.INSTALLED : Presence.UNKNOWN,
        installed_version: installed,
        required_version: required,
        path: dotnetPath,
        source,
      },
    ];
  }

  override async inspectDependencies(root: string): Promise<DependencyInfo | null> {
    // globAny already de-duplicates, so this list plays the role of the
    // Python `set(projects)` used for the count.
    const projects = globAny(root, PROJECT_GLOBS, MAX_PROJECT_MATCHES);
    if (projects.length === 0) {
      return null;
    }
    return {
      ecosystem: 'dotnet',
      manager: 'nuget',
      lockfile: null,
      lockfile_present: false,
      manifest_present: true,
      installed: Presence.UNKNOWN,
      dependency_count: null,
      dev_dependency_count: null,
      notes: [`${projects.length} project file(s) found`],
    };
  }

  override async inspectTests(root: string): Promise<string[]> {
    return this.detect(root) ? ['dotnet test'] : [];
  }

  override async inspectBuildCommands(root: string): Promise<string[]> {
    return this.detect(root) ? ['dotnet build'] : [];
  }

  override async healthChecks(_root: string, runtimes: RuntimeInfo[]): Promise<HealthIssue[]> {
    const issues: HealthIssue[] = [];
    for (const rt of runtimes) {
      if (rt.name === 'dotnet' && rt.presence === Presence.NOT_INSTALLED) {
        issues.push({
          severity: Severity.CRITICAL,
          code: 'dotnet.not_installed',
          title: '.NET SDK not found',
          message: 'No `dotnet` executable was found on PATH.',
          evidence: ['which dotnet -> not found'],
          recommendation:
            'Install the .NET SDK from https://dotnet.microsoft.com/download.',
          confidence: null,
        });
      }
    }
    return issues;
  }
}
